File: src/main/java/bad/robot/radiate/teamcity/JsonProjectUnmarshaller.java
package bad.robot.radiate.teamcity;

import bad.robot.http.HttpResponse;
import bad.robot.radiate.Unmarshaller;
import com.google.gson.Gson;

class JsonProjectUnmarshaller implements Unmarshaller<HttpResponse, Project> {

    @Override
    public Project unmarshall(HttpResponse response) {
        Project project = new Gson().fromJson(new JsonResponse(response).body(), FullProject.class);
        if (!project.iterator().hasNext())
            return new Project(project.getId(), project.getName(), project.getHref());
        return project;
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/JsonProjectUnmarshaller.scala
package bad.robot.radiate.teamcity

import argonaut.Argonaut._
import simplehttp.HttpResponse
import bad.robot.radiate.Unmarshaller
import bad.robot.radiate.Error
import bad.robot.radiate.ParseError
import scalaz.syntax.std.either._

import scalaz.\/

class JsonProjectUnmarshaller extends Unmarshaller[HttpResponse, Project] {
  def unmarshall(response: HttpResponse): Error \/ Project = {
    val json = new JsonResponse(response).body
    json.decodeEither[Project].disjunction.leftMap(ParseError)
  }
}